Mathlib/Combinatorics/SimpleGraph/Turan.lean

```diff
@@ -0,0 +1,92 @@
+* `turanGraph n r`: the canonical `r + 1`-cliquefree Turán graph on `n` vertices.
+
+## TODO
+
+* Port the rest of Turán's theorem from https://github.com/leanprover-community/mathlib4/pull/9317
+-/
+
+
+open Finset
+
+namespace SimpleGraph
+
+variable {V : Type*} [Fintype V] [DecidableEq V]
+variable (G : SimpleGraph V) [DecidableRel G.Adj]
+
+/-- An `r + 1`-cliquefree graph is `r`-Turán-maximal if any other `r + 1`-cliquefree graph on
+the same vertex set has the same or fewer number of edges. -/
+def IsTuranMaximal (r : ℕ) : Prop :=
+  G.CliqueFree (r + 1) ∧ ∀ (H : SimpleGraph V) [DecidableRel H.Adj],
+    H.CliqueFree (r + 1) → H.edgeFinset.card ≤ G.edgeFinset.card
+
+/-- The canonical `r + 1`-cliquefree Turán graph on `n` vertices. -/
+def turanGraph (n r : ℕ) : SimpleGraph (Fin n) where Adj v w := v % r ≠ w % r
+
+/-- `turanGraph n 0` is complete. Since the only 1-cliquefree graph is the empty graph,
+this shows that `0 < r` is needed in the statement of Turán's theorem. -/
+theorem turanGraph_zero_eq_top {n : ℕ} : turanGraph n 0 = ⊤ := by
+  ext a b
+  simp_rw [turanGraph, top_adj, Nat.mod_zero, not_iff_not, Fin.val_inj]
+
+variable {n r : ℕ}
+
+instance : DecidableRel (turanGraph n r).Adj := by dsimp only [turanGraph]; infer_instance
+
+/-- For `n ≤ r`, `turanGraph n r` is complete. -/
+theorem turanGraph_eq_top_of_le {n r} (h : n ≤ r) : turanGraph n r = ⊤ := by
+  ext a b
+  simp_rw [turanGraph, top_adj, Nat.mod_eq_of_lt (a.2.trans_le h),
+    Nat.mod_eq_of_lt (b.2.trans_le h), not_iff_not, Fin.val_inj]
+
+variable (hr : 0 < r)
+
+theorem turanGraph_cliqueFree : (turanGraph n r).CliqueFree (r + 1) := by
+  rw [cliqueFree_iff]
+  by_contra h
+  rw [not_isEmpty_iff] at h
+  obtain ⟨f, ha⟩ := h
+  simp only [turanGraph, top_adj] at ha
+  obtain ⟨x, y, d, c⟩ := Fintype.exists_ne_map_eq_of_card_lt (fun x : Fin (r + 1) ↦
+    (⟨(f x).1 % r, Nat.mod_lt _ hr⟩ : Fin r)) (by simp)
+  simp only [Fin.mk.injEq] at c
+  exact absurd c ((@ha x y).mpr d)
+
+/-- For `n ≤ r` and `0 < r`, `turanGraph n r` is Turán-maximal. -/
+theorem isTuranMaximal_turanGraph_of_le (h : n ≤ r) : (turanGraph n r).IsTuranMaximal r := by
+  refine' ⟨turanGraph_cliqueFree hr, _⟩
+  intro H _ _
+  exact card_le_card (edgeFinset_mono ((turanGraph_eq_top_of_le h).symm ▸ le_top (a := H)))
+
+/-- An `r + 1`-cliquefree Turán-maximal graph is _not_ `r`-cliquefree
+if it can accommodate such a clique. -/
+theorem not_cliqueFree_of_isTuranMaximal (hn : r ≤ Fintype.card V) (hx : IsTuranMaximal G r) :
+    ¬G.CliqueFree r := by
+  by_contra cf
+  obtain ⟨K, ⟨_, cK⟩⟩ := exists_smaller_set (univ : Finset V) r hn
+  have := (G.isNClique_iff).not.mp (cf K)
+  simp_rw [cK, and_true, IsClique, Set.Pairwise, not_forall] at this
+  obtain ⟨a, _, b, _, ne, na⟩ := this
+  have nhx : ¬G.IsTuranMaximal r := by
+    simp_rw [IsTuranMaximal, hx.1, true_and]; push_neg
+    use G ⊔ edge a b, inferInstance, cf.sup_edge a b
+    convert Nat.lt.base _
+    convert G.card_edgeFinset_sup_edge na ne
```

Comment:
Is there a nicer way to do these steps? `rw [G.card_edgeFinset_sup_edge na ne, Nat.lt.base]` doesn't work because the inner instances corresponding to the `Sup` don't match.

Reply:
I fixed it (by generalising the fintype assumptions appearing on the LHS of earlier lemmas) and then found a golf which made the fix redundant. Here's the fixed ungolfed version for your instruction:
```
theorem not_cliqueFree_of_isTuranMaximal (hn : r ≤ Fintype.card V) (hG : IsTuranMaximal G r) :
    ¬G.CliqueFree r := by
  by_contra cf
  obtain ⟨K, _, rfl⟩ := exists_smaller_set (univ : Finset V) r hn
  have := (cf K)
  rw [isNClique_iff] at this
  have := (G.isNClique_iff).not.mp (cf K)
  simp_rw [and_true, IsClique, Set.Pairwise, not_forall] at this
  obtain ⟨a, _, b, _, ne, na⟩ := this
  refine not_not_intro hG ?_
  simp_rw [IsTuranMaximal, hG.1, true_and]; push_neg
  use G ⊔ edge a b, inferInstance, cf.sup_edge a b
  rw [card_edgeFinset_sup_edge]
  exacts [lt_add_one _, na, ne]
```